src/jdataview.js

```diff
@@ -1,833 +1,471 @@
-		return toArray ? arrayFrom(result) : result;
-	},
+		const result = new Uint8Array(this.buffer, byteOffset, length);
 
-	_setBytes: function (byteOffset, bytes, littleEndian) {
-		var length = bytes.length;
-
-		// needed for Opera
-		if (length === 0) {
-			return;
-		}
+		return littleEndian || length <= 1
+			? result
+			: Array.from(result).reverse();
+	}
 
-		littleEndian = defined(littleEndian, this._littleEndian);
-		byteOffset = defined(byteOffset, this._offset);
+	#setBytes(
+		byteOffset = this.#bytePointer,
+		bytes,
+		littleEndian = this.littleEndian
+	) {
+		const length = bytes.length;
 
-		this._checkBounds(byteOffset, length);
+		this.#checkBounds(byteOffset, length);
 
 		if (!littleEndian && length > 1) {
-			bytes = arrayFrom(bytes, true).reverse();
+			bytes = Array.from(bytes).reverse();
 		}
 
 		byteOffset += this.byteOffset;
 
-		if (this._isArrayBuffer) {
-			new Uint8Array(this.buffer, byteOffset, length).set(bytes);
-		}
-		else {
-			if (NODE && this._isNodeBuffer) {
-				new Buffer(bytes).copy(this.buffer, byteOffset);
-			} else {
-				for (var i = 0; i < length; i++) {
-					this.buffer[byteOffset + i] = bytes[i];
-				}
-			}
-		}
-
-		this._offset = byteOffset - this.byteOffset + length;
-	},
+		new Uint8Array(this.buffer, byteOffset, length).set(bytes);
 
-	setBytes: function (byteOffset, bytes, littleEndian) {
-		this._setBytes(byteOffset, bytes, defined(littleEndian, true));
-	},
+		this.#bytePointer = byteOffset - this.byteOffset + length;
+	}
 
-	getString: function (byteLength, byteOffset, encoding) {
-		if (NODE && this._isNodeBuffer) {
-			byteOffset = defined(byteOffset, this._offset);
-			byteLength = defined(byteLength, this.byteLength - byteOffset);
+	/**
+	 * Get raw bytes. If length is undefined, it will go to the end of the buffer.
+	 */
+	getBytes(length, byteOffset, littleEndian) {
+		return this.#getBytes(length, byteOffset, littleEndian ?? true);
+	}
 
-			this._checkBounds(byteOffset, byteLength);
+	/**
+	 * Directly set raw bytes at `byteOffset` or the current pointer.
+	 */
+	setBytes(byteOffset, bytes, littleEndian) {
+		this.#setBytes(byteOffset, bytes, littleEndian ?? true);
+	}
 
-			this._offset = byteOffset + byteLength;
-			return this.buffer.toString(encoding || 'binary', this.byteOffset + byteOffset, this.byteOffset + this._offset);
-		}
-		var bytes = this._getBytes(byteLength, byteOffset, true);
+	/**
+	 * Read a string using the specified encoding, or binary if unspecified
+	 */
+	getString(byteLength, byteOffset, encoding = "binary") {
+		const bytes = this.#getBytes(byteLength, byteOffset, true);
 		// backward-compatibility
-		encoding = encoding === 'utf8' ? 'utf-8' : (encoding || 'binary');
-		if (TextDecoder && encoding !== 'binary') {
-			return new TextDecoder(encoding).decode(this._isArrayBuffer ? bytes : new Uint8Array(bytes));
+		if (encoding !== "binary") {
+			return new TextDecoder(encoding).decode(bytes);
```

Comment:
Note: creating instances of TextEncoder / TextDecoder can be relatively slow. Create those once in some variable outside the class and reuse instead.

Reply:
Wouldn't we need to create one for each encoding then?

Reply:
Yeah but we only used to support UTF-8, I suppose we can keep it that way. TextEncoder doesn't support anything else either.